README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# namedropR <img src="man/figures/logo.png" align="right" width="150"/>

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/namedropR)](https://CRAN.R-project.org/package=namedropR)
[![R-CMD-check](https://github.com/nucleic-acid/namedropR/workflows/R-CMD-check/badge.svg)](https://github.com/nucleic-acid/namedropR/actions)
[![Total
Downloads](https://cranlogs.r-pkg.org/badges/grand-total/namedropR)](https://CRAN.R-project.org/package=namedropR)
[![Monthly
Downloads](https://cranlogs.r-pkg.org/badges/namedropR)](https://CRAN.R-project.org/package=namedropR)

<!-- badges: end -->

`namedropR` provides ‘visual citations’ containing the metadata of a
scientific paper and a ‘QR’ code.

## Installation

To install, simply run:

``` r
install.packages("namedropR")
```

You can install the development version of namedropR from
[GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("nucleic-acid/namedropR")
```

## Visual Citations

A ‘visual citation’ is a banner containing an article’s title, authors,
journal and year of a publication (s. example below).

<img src="man/figures/collaboration_2019_ApJL.png" alt="A visual citation pointing to a scientific paper. Scan with a QR code scanner to follow the URL." align="right" width="500" style="box-shadow: 8px 8px 6px grey;"/>

One might want to include this in a presentation

- to back up one’s claims,
- to drop an important name,
- to boast about a publication in a ‘high impact journal’.

*Some might even use this to genuinely **point the audience to good
resources** for further reading.*

On conferences, such banners are frequently displayed way too short for
the audience to actually comprehend them (and often in a bad
resolution). Creating visual citations requires manually taking a
screenshot and placing it on the slide.

<img src="man/figures/hawking_1973.png" alt="A compact visual citation pointing to a scientific paper. Scan with a QR code scanner to follow the URL." align="left" width="130" style="box-shadow: 5px 5px 6px grey;margin-right: 15px"/>

`namedropR` helps to generate visual citations conveniently (see below),
in high resolution and with a QR code. This allows the audience to
follow the reference *while you are talking about it*, instead of
looking it up in a reference list at the end of your talk.  
There is also a less intrusive, compact style, as seen on the left.

## Basic Usage

### Working with bibtex bibliographies

This package accepts bibliographic information as ‘BibTeX’ and
‘BibLaTeX’ references and includes a QR code pointing to the
[‘DOI’](https://www.doi.org). If the ‘DOI’ is not available in the
bibliography entry, but a ‘URL’ field instead, this is used. If neither
is given, the QR code points to a search call at
<https://scholar.google.com> with the available data as search terms.

``` r
bib_path <- "path/to/bibliography_file.bib"

# load the package
library('namedropR')

drop_name(bib_path, cite_key = "SomeAuthor2010")
```

`drop_name()` by default returns the file path, where the visual
citation was stored as character string. Within an Rmarkdown/HTML
presentation you can render the VC and directly pass the path to
`{htmltools}` to include the banner:

``` r
# PNG
htmltools::img(
  src = drop_name(
    bib_path,
    cite_key = "SomeAuthor2010",
    export_as = "png"
  )
)

# HTML (note the recommended option use_xaringan, see documentation for details)
htmltools::includeHTML(
  drop_name(
    bib_path,
    cite_key = "SomeAuthor2010",
    export_as = "html",
    use_xaringan = TRUE
  )
)
```

To **bulk-create** VCs to include in another kind of document /
presentation slides you can pass a vector of citation keys or pass no
keys at all to render some or all bibliography entries respectively:

``` r
drop_name(
  bib_path,
  cite_key = c("SomeAuthor2010", "SomeOther2011", "YetAnother2012"),
  export_as = "png",
  style = "clean"
)
#> renders the three specified banners

drop_name(
  bib_path,
  export_as = "png",
  style = "clean"
)
#> renders all entries within the specified bibliography file
```

### Pulling data from Crossref

If you don’t have a \*.bib file, but DOIs instead, you can pass them to
`drop_name_crossref()` which pulls the data from
[Crossref](https://www.crossref.org) and passes the downloaded data on
to `drop_name()`.

**TEMPORARY NOTE:** *Due to recent API changes in Crossref you need to
have {rcrossref} version \>= 1.2 installed.*

``` r
# DOIs can be named vectors...
drop_name_crossref(c(cite1 = "10.1126/science.169.3946.635", cite2 = "10.1111/joms.12670"))

# ...or unnamed vectors
drop_name_crossref(c("10.1126/science.169.3946.635", "10.1111/joms.12670"))
```

All options, that can be passed to `drop_name()`, can also be specified
in `drop_name_crossref()`, for example the following call pulls data for
two references from Crossref and renders them as PNG files with the
‘newspaper’ style:

``` r
drop_name_crossref(
  c(cite1 = "10.1126/science.169.3946.635", cite2 = "10.1111/joms.12670"),
  export_as = "png",
  style = "newspaper"
)
```

## Options

It is possible to specify output format (HTML or PNG), style, size of
the QR code and much more. See the vignettes for all options.

These are some frequently used options:

``` r
# create a visual citation as PNG with 'modern' design
drop_name(bib_path, cite_key = "SomeAuthor2010", export_as = "png", style = "modern")

# create a visual citation as HTML with 'compact' design
drop_name(bib_path, cite_key = "SomeAuthor2010", export_as = "html", style = "compact")

# set a smaller QR code size, default is 250 (pixels)
drop_name(bib_path, cite_key = "SomeAuthor2010", export_as = "png", qr_size = 150)

# set a wider text area for long titles or long author lists, default is 600 (pixels)
drop_name(bib_path, cite_key = "SomeAuthor2010", export_as = "png", vc_width = 800)
```

## Styles

These are the currently predefined styles:

| Name      | Example                                                                                                                                                                                                       |
|-----------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| modern    | <img src="man/figures/modern.png" alt="A visual citation pointing to a scientific paper. Scan with a QR code scanner to follow the URL." align="right" width="500" style="box-shadow: 4px 4px 3px grey;"/>    |
| modern_bw | <img src="man/figures/modern_bw.png" alt="A visual citation pointing to a scientific paper. Scan with a QR code scanner to follow the URL." align="right" width="500" style="box-shadow: 4px 4px 3px grey;"/> |
| classic   | <img src="man/figures/classic.png" alt="A visual citation pointing to a scientific paper. Scan with a QR code scanner to follow the URL." align="right" width="500" style="box-shadow: 4px 4px 3px grey;"/>   |
| fancy     | <img src="man/figures/fancy.png" alt="A visual citation pointing to a scientific paper. Scan with a QR code scanner to follow the URL." align="right" width="500" style="box-shadow: 4px 4px 3px grey;"/>     |
| clean     | <img src="man/figures/clean.png" alt="A visual citation pointing to a scientific paper. Scan with a QR code scanner to follow the URL." align="right" width="500" style="box-shadow: 4px 4px 3px grey;"/>     |
| newspaper | <img src="man/figures/newspaper.png" alt="A visual citation pointing to a scientific paper. Scan with a QR code scanner to follow the URL." align="right" width="500" style="box-shadow: 4px 4px 3px grey;"/> |
| compact   | <img src="man/figures/compact.png" alt="A visual citation pointing to a scientific paper. Scan with a QR code scanner to follow the URL." align="left" width="180" style="box-shadow: 4px 4px 3px grey;"/>    |

Styling is possible via these predefined designs or via custom ‘CSS’
e.g. to match the design of the HTML presentation like e.g. ‘xaringan’
or ‘revealJS’ (see the vignette for more options).

## How to contribute

If you have ideas for Visual Citation styles and think, everyone should
be able to use them, feel free to [open an
issue](https://github.com/nucleic-acid/namedropR/issues) and describe
the style or provide the CSS code, if you have one already.

If you want to speed things up, you can fork the project to create a
pull request. Please also open an issue, so we can have a discussion, if
needed. See the
[readme](https://github.com/nucleic-acid/namedropR/blob/main/inst/styles/readme.md)
for details.

## Code of Conduct

Please note that the namedropR project is released with a [Contributor
Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.

## Notes

- This package is not intended as replacement for proper reference
  manager packages, but a tool to enrich scientific presentation slides.
  Hence the functionality is quite limited to this specific use case.
- previous versions (before v2.0) allowed passing
  `RefManageR::BibEntry()` objects. This option was removed to eliminate
  dependency on the orphaned `{bibtex}` package. This might break code
  for users of initial releases of the package, but was needed to allow
  for CRAN submission.


NEWS.md
# namedropR (development version)

* Minor syntax changes to comply with changes in {dplyr}

# namedropR 2.4.1

* Solves issue of failing tests in CRAN checks.

# namedropR 2.4.0

* Adds drop_name_crossref() function (thanks to Lukas Wallrich)
* Adds the ability to turn a QR code into a hyperlink (thanks to Matt Warkentin)
* Rebuilds documentation using latest roxygen2 version to adhere to HTML5 standards

# namedropR 2.3.3

* Fixes critical error previously missed by unit tests when working with file paths. This broke the basic functionality on some systems when working outside of projects.
* Removes dependency on the {here} package

# namedropR 2.3.2

* Fixes failing test due to tempdir access problems on win devel.
* Changes git install snippet in README.md to use {remotes} instead of {devtools}

# namedropR 2.3.1

* Adds documentation / vignette for new styling options
* changes lazy data loading to false (no relevant datasets are included)
* tested with R 4.2

namedropR 2.3.0

* Adds option to specify QR code color.
* Adds several options to adjust stylistic elements like font, color, size and weight for title, journal and authors.
* Adjusts predefined styles to allow for custom style editing.

# namedropR 2.2.3

* Fixes inconsistent documentation and presentation in the Readme.
* Fixes style options that resulted in sub-optimal readability of the rendered citations.
* Adds sample render images for the predefined styles in the readme.
* Slight adjustment to the DESCRIPTION title.

# namedropR 2.2.2

* Adds a {pkgdown} website to the repository to make the documentation more accessible.
* Adds a separate vignette for use in Rmarkdown environments.

# namedropR 2.2.1

* Allows users to also specify the width of the visual citation, as larger QR codes compressed the remaining text area.
* Adds two new styles to chose from.

# namedropR 2.2.0

* Styles are now stored in a separate csv. This allows for easier addition of styles.
* Allows user specification of QR code size.

# namedropR 2.1.1

* Removes examples from not-exported functions to fulfill CRAN reviewer comment
* Adds and improves test coverage for additional input situations.
* Minor style adjustments for compact and clean style.

# namedropR 2.1.0

* Implements 'compact' output style. See update vignette for usage.

# namedropR 2.0.2

* Slight adjustment to the 'modern' style (journal titles not capitalized anymore)
* By default curly braces are removed from titles and journal names. This can be prevented by setting clean_strings to FALSE

# namedropR 2.0.1

* Minor fixes to prepare pull request to main
* Style fixes of code in preparation for release

# namedropR 2.0.0

* Major code changes to eliminate dependency on the orphaned {bibtex} package to allow for CRAN submission.
* On the down side the package cannot work with BibEntry objects anymore, so the update might break existing user-code. Due to this the version number was increased as a 'major release'. 
* As new function, bulk handling is now available.
* The update furthermore improves overall stability and test coverage.

# namedropR 1.0.2

* Minor adjustments to unit tests to avoid build errors

# namedropR 1.0.1

* Minor adjustments to unit tests to avoid build errors

# namedropR 1.0.0

* Feature complete first version for CRAN submission.
* Improves unit testing.

# namedropR 0.10.1

* Corrects a typo in the example in the README.

# namedropR 0.10.0

* Added a vignette and proper README

# namedropR 0.9.4.9000

* Added a `NEWS.md` file to track changes to the package.


R/drop_name.R
#' @title drop_name
#'
#' @description Extracts metadata from a .bib file and exports the visual citation in the specified format.
#'
#' @param bib Accepts one of the following:
#' 1) A data.frame or tibble containing the columns YEAR, JOURNAL, AUTHOR, TITLE, BIBTEXKEY (all mandatory) and DOI, URL (optional).
#' 2) A file path to a bibliography file in BibTeX/BibLaTeX format (usually *.bib file).
#' 3) A string containing one or more BibTeX citations (must start with `@`)
#' @param cite_key If given, either a character string or a vector of strings are accepted.
#' Specifies the reference items within the bibliography for which visual citations should be created.
#' If no key is specified, a visual citation is created for ALL reference items within the bibliography.
#' In other words, either one, many or no BibTeX citation keys can be specified.
#' @param export_as A string specifying the desired output format. For now supports PNG and HTML.
#' Use "html" to include the 'bare' taglist (recommended for inclusion in Rmarkdown documents) or "html_full" to write a standalone .html file including <head> etc.
#' The PNG is a screenshot of the rendered HTML via the 'webshot' package. The filename represents this two step approach on purpose.
#' For webshot you need to install phantomJS once (see 'webshot' documentation).
#' @param to_clipboard Should visual citation be copied to clipboard?
#' @param output_dir A string specifying the relative path, where the rendered output files should be stored.
#' @param max_authors Integer number of maximum authors to print. If the number of authors exceeds this, the list is cropped accordingly.
#' @param include_qr Character string specifying the way the QR code should be included or if no QR code should be included.
#' 'embed' results in a stand alone <img> tag within the HTML object, other options are ignored for the time being.
#' 'link' (default) creates a PNG of the QR code and stores it in a subfolder of the HTML file's location. The HTML <img> tag links to this file then.
#' 'link_svg' creates a SVG of the QR code and stores it in a subfolder of the HTML file's location. The HTML <img> tag links to this file then.
#' 'none' creates no QR code.
#' @param style A string specifying the desired style for the visual citation. Possible values are:
#' "modern", "classic", "clean", "fancy", "newspaper", "compact" and "none".
#' If "compact" is given, the rendered VC contains
#' only the last name of the first author and the publication year, next to the QR code.
#' If "none" is given, the returned html can use a custom css file provided by the user.
#' This custom CSS file must specify styles for <div> classes "top-row", "title-row" and "author-row". (see vignette)
#' @param path_absolute Boolean to specify, whether the returned output path is a relative path or an absolute path.
#' @param use_xaringan Boolean to specify if an HTML output is intended to be included in an HTML presentation (like e.g. xaringan) or not.
#' When including the visual citation via htmltools::includeHTML(), the QR code needs to be in a subfolder
#' relative to the rendered presentation, not relative to the visual citation.
#' @param clean_strings Removes curly braces {} from titles and journal names, as they are often present in
#' BibTeX strings, but not needed for the rendering. TRUE by default, but can be set to FALSE, if the {} are needed.
#' @param qr_size Specifies the height/width of the rendered QR code in px. Default: 250px, minimum: 150px. Ignored for SVG output.
#' @param qr_color Specifies the foreground color of the QR code as hex-string, e.g. "#00FF00"; default is black: "#000000".
#' @param qr_hyperlink Logical. Should the QR code be a hyperlink?
#' @param vc_width Specifies the width of the text part of the visual citation in px.
#' This can be adjusted to accommodate e.g. untypically long or short titles. Default: 600px
#' @param ... Allows for custom style arguments to override predefined styles. Supported are: author_size, author_font,
#' author_weight, author_color, title_size, title_font, title_weight, title_color,
#' journal_size, journal_font, journal_weight, journal_color. Fonts need to be installed on the system.
#' @return A character string with the file path to the created visual citation in the specified output format.
#'
#' @examples
#'
#' # create sample data
#' \dontrun{
#' bib_tbl <- dplyr::tribble(
#'   ~TITLE, ~AUTHOR, ~JOURNAL, ~BIBTEXKEY, ~YEAR,
#'   "Some title", c("Alice", "Bob", "Charlie"),
#'   "Journal of Unnecessary R Packages",
#'   "Alice2022", "2022"
#' )
#'
#' # create visual citation
#' drop_name(
#'   bib = bib_tbl,
#'   cite_key = "Alice2022",
#'   export_as = "png",
#'   max_authors = 2,
#'   style = "clean",
#'   output_dir = "visual_citations",
#'   author_color = "#FF0000",
#'   author_weight = "normal",
#'   author_size = "12pt",
#'   author_font = "Roboto",
#'   title_color = "#00FF00",
#'   title_weight = "bold",
#'   title_size = "2.5rem",
#'   title_font = "Playfair Display",
#'   journal_color = "#0000FF",
#'   journal_weight = "bold",
#'   journal_size = "8pt",
#'   journal_font = "Fira Sans",
#'   qr_size = 150,
#'   qr_color = "#AAAAAA"
#' )
#' }
#'
#' @export
#' @import bib2df
#' @import dplyr
#' @importFrom htmltools tags save_html
#' @importFrom lubridate year ymd
#' @importFrom webshot webshot
#' @importFrom stringr str_detect str_remove_all str_trunc str_replace_all

drop_name <- function(bib, cite_key,
                      output_dir = "visual_citations",
                      export_as = "html",
                      to_clipboard = FALSE,
                      max_authors = 3,
                      include_qr = "link",
                      qr_size = 250,
                      qr_color = "#000000",
                      qr_hyperlink = FALSE,
                      vc_width = 600,
                      style = "modern",
                      path_absolute = FALSE,
                      use_xaringan = FALSE,
                      clean_strings = TRUE,
                      ...) {

  style_args <- list(...)

  # CHECK other ARGUMENTS
  stopifnot(is.character(output_dir))
  stopifnot(is.character(export_as))
  stopifnot(export_as %in% c("html", "html_full", "png"))
  stopifnot(is.logical(to_clipboard))
  stopifnot(is.numeric(max_authors))
  stopifnot(max_authors >= 0)
  stopifnot(is.character(include_qr))
  stopifnot(include_qr %in% c("embed", "link", "link_svg", "none"))
  stopifnot(is.numeric(qr_size))
  if (qr_size < 150) {
    warning("QR size must be at least 150px. This will now be set as size.")
    qr_size <- 150
  }
  stopifnot(is.character(qr_color))
  if (!stringr::str_detect(qr_color, pattern = "^#[A-Fa-f\\d]{6}$")) {
    warning("QR color code does not match requirements. This needs to be a 6 digit hex code, like '#AA12FC'. Black will be used instead.")
    qr_color <- "#000000"
  }

  stopifnot(vc_width > 0)
  if (vc_width < 200) {
    message("You specified a fairly small 'vc_width'. If you want to have a compact visual citation, the 'compact' style might be handy.")
  }
  stopifnot(is.character(style))
  # style content is not further checked, as unknown styles will be handled as "none" in get_css_styles()
  stopifnot(is.logical(path_absolute))
  stopifnot(is.logical(use_xaringan))
  stopifnot(is.logical(qr_hyperlink))
  stopifnot(is.logical(clean_strings))

  # READ AND CHECK BIB FILE or BIBENTRY

  if (missing(bib)) {
    stop("No bibliography object or file path provided. Please check arguments.")
  }

  if (is.data.frame(bib)) {
    bib_data <- bib
  } else if (is.character(bib)) {
    bib_file <- NULL #Track whether new file was created
    if(substr(bib, 1, 1) == "@") {
      bib <- gsub("(?<=\\w{1})\\=\\{", " \\= \\{", bib, perl = TRUE) # To deal with Google Scholar parsing issue - https://github.com/nucleic-acid/namedropR/issues/49
      bib_file <- tempfile(fileext = ".bib")
      writeLines(bib, bib_file)
      writeLines(bib, "test.bib")
      bib <- bib_file
    }
    if (file.exists(bib)) {
      bib_data <- suppressMessages(suppressWarnings(bib2df::bib2df(file = bib)))
      if ("YEAR" %in% colnames(bib_data)) {
        if (!is.character(bib_data$YEAR)) {
          bib_data$YEAR <- as.character(bib_data$YEAR)
          message("Years coerced to string format.")
        }
      }
      if(!is.null(bib_file)) {
        file.remove(bib_file)
      }
      message("Bibliography file successfully read.")
    } else {
      stop("BibTeX file not found. Check file path or pass a suitable data.frame/tibble to the function.")
    }
  } else {
    stop("Inappropriate type of bibliography provided, please pass a data.frame, tibble or file path to a *.bib file.")
  }

  # check if data.frame is empty
  if (nrow(bib_data) == 0) {
    stop("Bibliography is empty.")
  }

  # provide COMPATIBILITY for biblatex fields JOURNALTITLE and DATE
  if ("DATE" %in% colnames(bib_data)) {
    parsed_year <- suppressWarnings(lubridate::year(lubridate::ymd(bib_data$DATE, truncated = 2)))
    # if (any(is.na(parsed_year))) {
    #   warning("Having difficulties extracting the year from at least one of the references' date field. Resulting output might not be as expected. Please make sure your dates comply with BibLaTeX standards 'YYYY', 'YYYY-MM' or 'YYYY-MM-DD'.")
    # }
    if ("YEAR" %in% colnames(bib_data)) {
      bib_data <- bib_data %>%
        dplyr::mutate(
          YEAR = ifelse(
            is.na(.data$YEAR),
            parsed_year,
            .data$YEAR
          )
        )
    } else {
      bib_data$YEAR <- parsed_year
    }
    # At this point no further error catching is implemented. As of current knowledge lubridate::year()
    # returns NA, if parsing the string fails. In drop_html() these NAs will be replaced with an empty string.
    # message("One or more references had BibLaTeX field 'DATE' and were transformed to 'YEAR'.")
  }


  if ("JOURNALTITLE" %in% colnames(bib_data)) {
    if ("JOURNAL" %in% colnames(bib_data)) {
      bib_data <- bib_data %>%
        dplyr::mutate(JOURNAL = ifelse(is.na(.data$JOURNAL), .data$JOURNALTITLE, .data$JOURNAL))
    } else {
      bib_data <- bib_data %>%
        dplyr::rename(JOURNAL = "JOURNALTITLE")
    }
    # message("One or more references had BibLaTeX field 'JOURNALTITLE' and were transformed to 'JOURNAL'.")
  }

  # print(bib_data)

  # check for required columns
  required_cols <- c("YEAR", "AUTHOR", "TITLE", "JOURNAL", "BIBTEXKEY")
  if (!all(required_cols %in% colnames(bib_data))) {
    stop("Required data.frame columns are 'YEAR' (or 'DATE'), 'AUTHOR', 'TITLE', 'JOURNAL', 'BIBTEXKEY' (all uppercase). At least one is missing or misspelled.")
  }

  # CLEAN STRINGS

  if (clean_strings) {
    bib_data <- bib_data %>%
      dplyr::mutate(
        TITLE = gsub("\\{|\\}", "", .data$TITLE),
        JOURNAL = gsub("\\{|\\}", "", .data$JOURNAL)
      )
  }

  # COMPOSE URL FOR QR CODE
  # If bib_data has a DOI column, use this as this first.
  # Next option is using the URL in the bib_data
  # If neither is available (or missing in some rows), a search query for Google Scholar will be passed to the QR code.
  # This will later be passed on to the HTML rendering function.

  # initialize empty QR column
  bib_data$QR <- NA

  # if DOI is available adapt this to the QR column
  if ("DOI" %in% colnames(bib_data)) {
    bib_data <- bib_data %>%
      dplyr::mutate(
        QR = ifelse(
          is.na(.data$DOI),
          NA,
          paste0("https://doi.org/", .data$DOI)
        )
      )
  }

  # if NO DOI but a URL is available adapt this to the QR column
  if ("URL" %in% colnames(bib_data)) {
    bib_data <- bib_data %>%
      dplyr::mutate(
        QR = ifelse(
          is.na(.data$QR) & !is.na(.data$URL),
          .data$URL,
          .data$QR
        )
      )
  }

  # the remaining URLs will be created below, after condensing the authors lists

  # select required columns only and add missing BIBTEXKEYs:
  clean_bib <- dplyr::select(bib_data, "YEAR", "AUTHOR", "JOURNAL", "TITLE", "BIBTEXKEY", "QR") %>%
    dplyr::mutate(
      `BIBTEXKEY` = ifelse(
        is.na(.data$BIBTEXKEY),
        stringr::str_remove_all(paste0(stringr::str_trunc(.data$TITLE, 10, ellipsis = ""), stringr::str_trunc(.data$JOURNAL, 10, ellipsis = ""), .data$YEAR), " "),
        .data$BIBTEXKEY
      )
    )

  # check for duplicate cite_keys
  if (any(dplyr::count(bib_data, .data$BIBTEXKEY)$n > 1)) {
    warning("There are duplicated bibtex keys / citation keys in this bibliography. I'll try to still render all selected entries. Make sure to check, whether the visual citations are as expected.")
  }

  # Ensure BIBTEXKEY are unique.
  letters_blank <- c("", letters)
  clean_bib <- dplyr::ungroup(dplyr::mutate(dplyr::group_by(clean_bib, .data$BIBTEXKEY), `BIBTEXKEY_2` = stringr::str_replace_all(paste0(.data$BIBTEXKEY, letters_blank[1:dplyr::n()]), " ", "_")))


  # create output directory, if needed
  if (!dir.exists(output_dir)) {
    tryCatch(
      expr = dir.create(output_dir, recursive = TRUE),
      error = function(e) {
        message("Could not create the output directory. Check file permissions.")
        print(e)
      },
      warning = function(w) {
        message("Having difficulties creating the output directory:")
        print(w)
      }
    )
  }


  # CALL HTML RENDERING FUNCTION


  # handle different input options for bibtex keys
  if (missing(cite_key)) {
    n_bib <- nrow(clean_bib)
    message(paste0("No cite_key specified. Working through all possible ", n_bib, " entries in the bibliography."))
    work_list <- clean_bib
  } else if (!is.character(cite_key)) {
    stop("cite_key must be of type 'caracter'.")
  } else {
    n_bib <- length(cite_key)
    if (n_bib > 1) {
      message(paste0(n_bib, " cite_key(s) specified. Working through all of them."))
    }
    missing_keys <- cite_key[!(cite_key %in% clean_bib$BIBTEXKEY)]
    if (length(missing_keys) > 0) {
      warning(paste0("The following cite_key items were not found in the provided library: ", missing_keys))
    }
    work_list <- clean_bib %>%
      dplyr::filter(.data$BIBTEXKEY %in% cite_key)
  }

  if (nrow(work_list) == 0) {
    warning("No reference matches the given cite_keys. Please check that citation key(s) are correct.")
    return("No Visual Citation created.")
  }

  authors_collapsed <- sapply(
    work_list$AUTHOR, manage_authors,
    max_authors = max_authors,
    style = style
  )


  work_list <- work_list %>%
    dplyr::mutate(
      authors_collapsed = authors_collapsed, # add collapsed authors
      `BIBTEXKEY` = .data$BIBTEXKEY_2 # replace duplicated keys with deduplicated keys
    )

  # now create QR URLs for remaining missing entries with short author list
  work_list <- work_list %>%
    dplyr::mutate(
      QR = ifelse(
        is.na(.data$QR),
        paste0(
          "https://scholar.google.com/scholar?as_q=",
          sub("[];<].*", "", .data$authors_collapsed), "+",
          .data$JOURNAL, "+",
          .data$YEAR
        ),
        .data$QR
      )
    )

  vcs <- apply(
    work_list,
    1,
    drop_html,
    include_qr = include_qr,
    style = style,
    output_dir = output_dir,
    # if PNG is desired output format, the relative filepath must not
    # be adapted for later inclusion of the HTML. Therefore use_xaringan
    # must be ignored, as otherwise the QR will not be included properly
    use_xaringan = ifelse(export_as == "png", FALSE, use_xaringan),
    qr_size = qr_size,
    qr_color = qr_color,
    qr_hyperlink = qr_hyperlink,
    vc_width = vc_width,
    style_args = style_args
  )

  work_list <- work_list %>%
    dplyr::mutate(
      vcs = vcs
    )

  file_paths <- apply(work_list, 1,
                      FUN = write_vc,
                      output_dir = output_dir,
                      path_absolute = path_absolute,
                      export_as = export_as,
                      to_clipboard = to_clipboard
  )

  return(file_paths)
}
